tests/tests/loopback.rs

```diff
@@ -0,0 +1,306 @@
+use anyhow::Result;
+use opteadm::OpteAdm;
+use oxide_vpc::api::{
+    AddRouterEntryReq, Address, BoundaryServices, Direction, Filters,
+    FirewallAction, FirewallRule, IpCfg, IpCidr, Ipv4Cfg, PhysNet, Ports,
+    RouterTarget, SNat4Cfg, SetVirt2PhysReq, Vni, VpcCfg,
+};
+use std::env;
+use std::process::Command;
+use ztest::*;
+
+/// This is a wrapper around the ztest::Zone object that encapsulates common
+/// logic needed for running the OPTE tests zones used in this test suite.
+struct OpteZone<'a> {
+    _zfs: &'a Zfs,
+    zone: Zone,
+}
+
+impl<'a> OpteZone<'a> {
+    /// Create a new zone with the given name, underlying zfs instance and set
+    /// of interfaces. In illumos parlance, the interfaces are data link
+    /// devices.
+    fn new(name: &str, zfs: &'a Zfs, ifx: &[&'a str]) -> Result<Self> {
+        // If we're running in CI, use the omicron1 zone, because that's whats
+        // there. If this is running locally, use a sparse zone which is much
+        // easier to set up on a regular Helios dev box.
+        let brand = match env::var("BUILDOMAT_JOB_ID") {
+            Ok(_) => "omicron1",
+            _ => "sparse",
+        };
+        let zone = Zone::new(name, brand, zfs, ifx)?;
+        Ok(Self { _zfs: zfs, zone })
+    }
+
+    /// Wait for the network to come up, then set up the overlay network.
+    fn setup(&self, devname: &str, addr: String) -> Result<()> {
+        self.zone.wait_for_network()?;
+        self.zone
+            .zexec(&format!("ipadm create-addr -t -T dhcp {}/test", devname))?;
+        self.zone.zexec(&format!("route add -iface 10.0.0.254 {}", addr))?;
+        self.zone.zexec("route add 10.0.0.0/24 10.0.0.254")?;
+        Ok(())
+    }
+}
+
+/// This is a wrapper around an OPTE port configuration. It provides a number of
+/// methods to streamline test-suite-specific OPTE configuration. It also
+/// deletes itself from the underlying OPTE kernel module state when dropped.
+struct OptePort {
+    name: String,
+    cfg: VpcCfg,
+}
+
+impl OptePort {
+    /// Create a new OPTE port with the provided `name`. The `private_ip` and
+    /// `guest_mac` parameters represent the ip/mac combo in a VM instance. The
+    /// `phys_ip` is the underlay address this port will use for sourcing
+    /// encapsulated traffic.
+    fn new(
+        name: &str,
+        private_ip: &str,
+        guest_mac: &str,
+        phys_ip: &str,
```

Comment:
Why not take appropriate types instead of strings?

Reply:
This is to capture all the `parse().unwrap()` boilerplate in one function.